libcontainer/setns_init_linux.go

```diff
@@ -49,6 +50,17 @@ func (l *linuxSetnsInit) Init() error {
 			}
 		}
 	}
+
+	// Clean the cache of RLIMIT_NOFILE in go runtime
+	// The problem originates from https://github.com/golang/go/commit/f5eef58e4381259cbd84b3f2074c79607fb5c821
+	rlimit := syscall.Rlimit{}
+	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit); err != nil {
+		return err
+	}
+	if err := syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rlimit); err != nil {
+		return err
+	}
```

Comment:
As I suggested in the earlier review, for correctness reasons, I think

```suggestion
	for _, rlimit := range l.config.Rlimits {
		if rlimit.Type == unix.RLIMIT_NOFILE {
			if err := unix.Setrlimit(rlimit.Type, &unix.Rlimit{
				Cur: rlimit.Soft,
				Max: rlimit.Hard,
			}); err != nil {
				return fmt.Errorf("failed to re-apply nofile rlimit: %w", err)
			}
		}
	}
```

makes more sense.

Reply:
Yes, the proposed implementation is better

Reply:
But running the unit test failed and I encountered an error:
=== RUN TestInitJoinNetworkAndUser
     exec_test.go:1543: unexpected error: unable to start container process: error during container init: failed to re-apply nofile rlimit: operation not permitte

My code just clears the cache in go runtime and has no meaning to the operating system.

Reply:
Just curious about 2 things:
1. I think the code provided by you and cyphar have the same effect, but I don't know why cyphar's code will be fail, could you please give some explainations? If yes, it will be appreciated.
2. If I understand right, there is no similar bug for `runc run`, this bug is only in `runc exec`? If yes, in `setns_init_linux`, I think we can use `procReady` msg to tell the parent process do `Setrlimit`, like what is doing in `standard_init_linux`.

Reply:
> 2\. like what is doing in `standard_init_linux`.

https://github.com/opencontainers/runc/blob/f9f8abf3102e47f2fc3aa45bcd134e957da7de66/libcontainer/process_linux.go#L775-L781